File: src/main/java/me/cjcrafter/armormechanics/ArmorMechanics.java
package me.cjcrafter.armormechanics;

import me.cjcrafter.armormechanics.listeners.*;
import me.cjcrafter.auto.UpdateChecker;
import me.cjcrafter.auto.UpdateInfo;
import me.deecaad.core.file.SerializeData;
import me.deecaad.core.file.SerializerException;
import me.deecaad.core.file.TaskChain;
import me.deecaad.core.utils.Debugger;
import me.deecaad.core.utils.FileUtil;
import me.deecaad.core.utils.LogLevel;
import me.deecaad.core.utils.ReflectionUtil;
import org.bstats.bukkit.Metrics;
import org.bstats.charts.SimplePie;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

public class ArmorMechanics extends JavaPlugin {

    public static ArmorMechanics INSTANCE;

    public Debugger debug;
    private Metrics metrics;
    private UpdateChecker update;

    public final Map<String, BonusEffect> effects = new HashMap<>();
    public final Map<String, ItemStack> armors = new HashMap<>();
    public final Map<String, ArmorSet> sets = new HashMap<>();


    @Override
    public void onLoad() {
        INSTANCE = this;

        int level = getConfig().getInt("Debug_Level", 2);
        boolean printTraces = getConfig().getBoolean("Print_Traces", false);
        debug = new Debugger(getLogger(), level, printTraces);

        if (ReflectionUtil.getMCVersion() < 13) {
            debug.error("  !!!!! ERROR !!!!!", "  !!!!! ERROR !!!!!", "  !!!!! ERROR !!!!!", "  Plugin only supports Minecraft 1.13 and higher");
            getServer().getPluginManager().disablePlugin(this);
            return;
        }
    }

    @Override
    public void onEnable() {

        reload();
        registerBStats();
        registerUpdateChecker();

        PluginManager pm = getServer().getPluginManager();
        pm.registerEvents(new ArmorEquipListener(), this);
        pm.registerEvents(new ArmorUpdateListener(), this);
        pm.registerEvents(new BlockPlaceListener(), this);
        pm.registerEvents(new DamageMechanicListener(), this);
        pm.registerEvents(new ImmunePotionCanceller(), this);
        pm.registerEvents(new PreventRemoveListener(), this);
        pm.registerEvents(new WeaponMechanicsDamageListener(), this);

        // Try to hook into MythicMobs, an error will be thrown if the user is
        // using any version below v5.0.0
        if (pm.getPlugin("MythicMobs") != null) {
            try {
                pm.registerEvents(new MythicMobsListener(), this);
            } catch (Throwable e) {
                debug.log(LogLevel.ERROR, "Could not hook into MythicMobs", e);
            }
        }

        Command.register();
    }

    public void reload() {

        // Write config from jar to datafolder
        if (!getDataFolder().exists() || getDataFolder().listFiles() == null || getDataFolder().listFiles().length == 0) {
            debug.info("Copying files from jar (This process may take up to 30 seconds during the first load!)");
            FileUtil.copyResourcesTo(getClassLoader().getResource("ArmorMechanics"), getDataFolder().toPath());
        }

        reloadConfig();

        // Clear old data
        effects.clear();
        armors.clear();
        sets.clear();

        // Serialize armor types
        File armorFile = new File(getDataFolder(), "Armor.yml");
        FileConfiguration armorConfig = YamlConfiguration.loadConfiguration(armorFile);

        for (String key : armorConfig.getKeys(false)) {
            ArmorSerializer serializer = new ArmorSerializer();
            SerializeData data = new SerializeData(serializer, armorFile, key, armorConfig);

            try {
                serializer.serialize(data);
            } catch (SerializerException e) {
                e.log(debug);
            }
        }

        File setFile = new File(getDataFolder(), "Set.yml");
        FileConfiguration setConfig = YamlConfiguration.loadConfiguration(setFile);

        for (String key : setConfig.getKeys(false)) {
            ArmorSet serializer = new ArmorSet();
            SerializeData data = new SerializeData(serializer, setFile, key, setConfig);

            try {
                serializer.serialize(data);
            } catch (SerializerException e) {
                e.log(debug);
            }
        }
    }

    private void registerBStats() {
        if (metrics != null) return;

        debug.debug("Registering bStats");

        // See https://bstats.org/plugin/bukkit/ArmorMechanics/15777. This is
        // the bStats plugin id used to track information.
        int id = 15777;

        this.metrics = new Metrics(this, id);

        metrics.addCustomChart(new SimplePie("registered_armors", () -> {
            int count = armors.size();

            if (count <= 10) {
                return "0-10";
            } else if (count <= 20) {
                return "11-20";
            } else if (count <= 30) {
                return "21-30";
            } else if (count <= 50) {
                return "31-50";
            } else if (count <= 100) {
                return "51-100";
            } else {
                return ">100";
            }
        }));

        metrics.addCustomChart(new SimplePie("registered_sets", () -> {
            int count = sets.size();

            if (count <= 2) {
                return "0-2";
            } else if (count <= 5) {
                return "3-5";
            } else if (count <= 10) {
                return "6-10";
            } else if (count <= 20) {
                return "11-20";
            } else if (count <= 50) {
                return "21-50";
            } else {
                return ">50";
            }
        }));
    }

    private void registerUpdateChecker() {
        update = new UpdateChecker(this, UpdateChecker.github("WeaponMechanics", "ArmorMechanics"));

        Listener listener = new Listener() {
            @EventHandler
            public void onJoin(PlayerJoinEvent event) {
                if (event.getPlayer().isOp()) {
                    new TaskChain(ArmorMechanics.this)
                            .thenRunAsync((callback) -> update.hasUpdate())
                            .thenRunSync((callback) -> {
                                UpdateInfo update = (UpdateInfo) callback;
                                if (callback != null)
                                    event.getPlayer().sendMessage(ChatColor.RED + "ArmorMechanics is out of date! " + update.current + " -> " + update.newest);

                                return null;
                            });
                }
            }
        };

        Bukkit.getPluginManager().registerEvents(listener, this);
    }
}

File: src/main/java/com/cjcrafter/armormechanics/ArmorMechanics.kt
package com.cjcrafter.armormechanics

import com.cjcrafter.armormechanics.commands.Command
import com.cjcrafter.armormechanics.listeners.ArmorEquipListener
import com.cjcrafter.armormechanics.listeners.ArmorUpdateListener
import com.cjcrafter.armormechanics.listeners.BlockPlaceListener
import com.cjcrafter.armormechanics.listeners.DamageMechanicListener
import com.cjcrafter.armormechanics.listeners.ImmunePotionCanceller
import com.cjcrafter.armormechanics.listeners.MythicMobsListener
import com.cjcrafter.armormechanics.listeners.PreventRemoveListener
import com.cjcrafter.armormechanics.listeners.WeaponMechanicsDamageListener
import me.deecaad.core.MechanicsPlugin
import me.deecaad.core.events.QueueSerializerEvent
import me.deecaad.core.file.Configuration
import me.deecaad.core.file.FastConfiguration
import me.deecaad.core.file.IValidator
import me.deecaad.core.file.JarInstancer
import me.deecaad.core.file.RootFileReader
import me.deecaad.core.file.SearchMode
import me.deecaad.core.file.SerializerInstancer
import org.bstats.charts.SimplePie
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import java.util.concurrent.Callable
import java.util.concurrent.CompletableFuture
import java.util.jar.JarFile

class ArmorMechanics : MechanicsPlugin(bStatsId = 15777) {

    var armorConfigurations: Configuration = FastConfiguration()
        private set
    var setConfigurations: Configuration = FastConfiguration()
        private set

    override fun onLoad() {
        INSTANCE = this
        super.onLoad()
    }

    override fun handleCommands(): CompletableFuture<Void> {
        Command.register()
        return super.handleCommands()
    }

    override fun handleConfigs(): CompletableFuture<Void> {
        // Look for all serializers/validators in the jar
        val jar = JarFile(file)
        val serializers = SerializerInstancer(jar).createAllInstances(classLoader, SearchMode.ON_DEMAND)
        val validators = JarInstancer(jar).createAllInstances(IValidator::class.java, classLoader, SearchMode.ON_DEMAND)

        val event = QueueSerializerEvent(this, dataFolder)
        event.addSerializers(serializers)
        event.addValidators(validators)
        server.pluginManager.callEvent(event)

        armorConfigurations = RootFileReader(this, Armor::class.java, "armors")
            .withSerializers(event.serializers)
            .withValidators(event.validators)
            .assertFiles()
            .read()
        debugger.info("Loaded ${armorConfigurations.keys(deep = false).size} armors")

        setConfigurations = RootFileReader(this, ArmorSet::class.java, "sets")
            .withSerializers(event.serializers)
            .withValidators(event.validators)
            .assertFiles()
            .read()
        debugger.info("Loaded ${setConfigurations.keys(deep = false).size} sets")

        return super.handleConfigs()
    }

    override fun handleListeners(): CompletableFuture<Void> {
        server.pluginManager.run {
            val plugin = this@ArmorMechanics
            registerEvents(ArmorEquipListener(), plugin)
            registerEvents(ArmorUpdateListener(), plugin)
            registerEvents(BlockPlaceListener(), plugin)
            registerEvents(DamageMechanicListener(), plugin)
            registerEvents(ImmunePotionCanceller(), plugin)
            registerEvents(PreventRemoveListener(), plugin)

            if (getPlugin("WeaponMechanics") != null) {
                registerEvents(WeaponMechanicsDamageListener(), plugin)
            }

            // Try to hook into MythicMobs, an error will be thrown if the user is
            // using any version below v5.0.0
            if (getPlugin("MythicMobs") != null) {
                try {
                    registerEvents(MythicMobsListener(), plugin)
                } catch (e: Throwable) {
                    debugger.severe("Could not hook into MythicMobs", e)
                }
            }

            // Automatically reload ArmorMechanics if WeaponMechanics reloads (only after 20 ticks to avoid
            // the first load from WM)
            foliaScheduler.global().runDelayed(Runnable {
                registerEvents(object : Listener {
                    @EventHandler
                    fun onQueue(event: QueueSerializerEvent) {
                        if ("WeaponMechanics" == event.sourceName) reload()
                    }
                }, plugin)
            }, 20L)
        }

        return super.handleListeners()
    }

    override fun handleMetrics(): CompletableFuture<Void> {
        metrics!!.addCustomChart(SimplePie("registered_armors", Callable {
            val count = armorConfigurations.keys(deep = false).size
            if (count <= 10) {
                return@Callable "0-10"
            } else if (count <= 20) {
                return@Callable "11-20"
            } else if (count <= 30) {
                return@Callable "21-30"
            } else if (count <= 50) {
                return@Callable "31-50"
            } else if (count <= 100) {
                return@Callable "51-100"
            } else {
                return@Callable ">100"
            }
        }))
        metrics!!.addCustomChart(SimplePie("registered_sets", Callable {
            val count = setConfigurations.keys(deep = false).size
            if (count <= 2) {
                return@Callable "0-2"
            } else if (count <= 5) {
                return@Callable "3-5"
            } else if (count <= 10) {
                return@Callable "6-10"
            } else if (count <= 20) {
                return@Callable "11-20"
            } else if (count <= 50) {
                return@Callable "21-50"
            } else {
                return@Callable ">50"
            }
        }))
        return super.handleMetrics()
    }

    companion object {
        private lateinit var INSTANCE: ArmorMechanics

        @JvmStatic
        fun getInstance(): ArmorMechanics {
            return INSTANCE
        }
    }
}
